chown() and chmod()

xsawyerx opened the pull request:
* The commands are added to strict mode.
* In chmod(), we try to issue the right warning, but the line number we're using is right above the line that *should* issue it. I think that's okay enough.
* chown() provides granular error if you send it t directory that does not exist vs. a file. chmod() doesn't seem to do that.
* If you call chown() or chmod() on mixed set of files (some mocked, some not), it will die.

```
    my $foo = Test::MockFile->file( '/foo/bar', '' );
    chmod 0644, '/foo/bar', '/bar/baz'; # dies
    chmod 0644, '/bar/baz';             # ok without strict mode
    chmod 0644, '/bar/baz';             # dies on strict mode
```

Resolves GH #15.

atoomic commented on lib/Test/MockFile.pm:
```diff
@@ -1712,6 +1714,120 @@ BEGIN {
+        # it's probably a mistake
+        my $mocked_count = grep ref, values %mocked_files;
+        if ( $mocked_count && $mocked_count != @files ) {
+            die "You called chown() on a mix of mocked and unmocked files - this is very likely a bug on your side";
```
maybe list the files?

atoomic replied:
or provide some context to know where it's coming from?
as most of the time it s going to be action at a distance

xsawyerx replied:
Good idea. Will do.

toddr commented on lib/Test/MockFile.pm:
```diff
@@ -1712,6 +1714,120 @@ BEGIN {
+
+            if ( !$mock->exists() ) {
+                if ( $mock->is_file() ) {
+                    $! = ENOENT;
```
does $! really clobber like this for a list of files? or does it fail on the first one?

xsawyerx replied:
I imagine the clobbering is certainly wrong. Either it should only clobber with errors (so a non-error would not obscure an error) or it shouldn't clobber at all. I'll check and revise. Thanks!

toddr commented on lib/Test/MockFile.pm:
```diff
@@ -1712,6 +1714,120 @@ BEGIN {
+                    $! = ENOENT;
+                }
+
+                if ( $mock->is_dir() ) {
```
elsif? make it clear that only one of these happens?

xsawyerx replied:
D'oh.

toddr commented on lib/Test/MockFile.pm:
```diff
@@ -1712,6 +1714,120 @@ BEGIN {
+                }
+
+                if ( $mock->is_dir() ) {
+                    $! = ENOTDIR;
```
If the path doesn't exist, it's always going to be the same error from perl's perspective, right?

```
$>perl -E'say chown (0, 0, qw{/nofile /notafile }); say $!+0; say "$!"'
0
2
No such file or directory
```

toddr replied:
2 seems to be `ENOENT`?

xsawyerx replied:
I was certain I tested it with `foo/` (ending with a forward slash) and it gave a different error but I can't replicate it, so I guess I saw it elsewhere and mistook it for this scenario. I'll correct this. Thanks.

xsawyerx commented:
Changes since last submission:

* Calling `chmod` and `chown` on mixed values will now output the filenames for both mocked and unmocked files. This is not applicable on strict mode which wouldn't allow unmocked anyway. Raised by @atoomic.
* Optimized the permission check on `chown` and if you hit `EPERM`, you don't even loop over files because there's no point.
* `$!` doesn't get clobbered. Raised by @toddr.
* Removed incorrect `ENODIR` when `ENOENT` is the only error. This also removes superfluous `if()` conditions. Raised by @toddr.
